heudiconv/utils.py

```diff
@@ -681,12 +684,123 @@ def strptime_micr(date_string: str, fmt: str) -> datetime:
+
+
+def datetime_utc_offset(datetime_obj: datetime.datetime, utc_offset: str) -> datetime.datetime:
+    """set the datetime's tzinfo by parsing an utc offset string"""
+    # https://dicom.innolitics.com/ciods/electromyogram/sop-common/00080201
+    extract_offset = re.match(r"([+\-]?)(\d{2})(\d{2})", utc_offset)
+    if extract_offset is None:
+        raise ValueError(f"utc offset {utc_offset} is not valid")
+    sign, hours, minutes = extract_offset.groups()
+    sign = -1 if sign == '-' else 1
+    hours, minutes = int(hours), int(minutes)
+    tzinfo = datetime.timezone(sign * datetime.timedelta(hours=hours, minutes=minutes))
+    return datetime_obj.replace(tzinfo=tzinfo)
+
+
+def strptime(datetime_string: str, fmts: list[str]) -> datetime.datetime:
+    r"""
+    Try datetime.strptime on a list of formats returning the first successful attempt.
+
+    Parameters
+    ----------
+    datetime_string: str
+      Datetime string to parse
+    fmts: list[str]
+      List of format strings
+    """
+    datetime_str = datetime_string.strip()
+    for fmt in fmts:
+        try:
+            return datetime.datetime.strptime(datetime_str, fmt)
+        except ValueError:
+            pass
+    raise ValueError(f"Unable to parse datetime string: {datetime_str}")
+
+
+def strptime_bids(datetime_string: str) -> datetime.datetime:
+    r"""
+    Create a datetime object from a bids datetime string.
+
+    Parameters
+    ----------
+    date_string: str
+      Datetime string to parse
+    """
+    # https://bids-specification.readthedocs.io/en/stable/common-principles.html#units
+    fmts = ["%Y-%m-%dT%H:%M:%S.%f%z", "%Y-%m-%dT%H:%M:%S%z", "%Y-%m-%dT%H:%M:%S.%f", "%Y-%m-%dT%H:%M:%S"]
+    datetime_obj = strptime(datetime_string, fmts)
+    return datetime_obj
+
+
+def strptime_dcm_da_tm(dcm_data: dcm.Dataset, da_tag: TagType, tm_tag: TagType) -> datetime.datetime:
+    r"""
+    Create a datetime object from a dicom DA tag and TM tag.
+
+    Parameters
+    ----------
+    dcm_data : dcm.Dataset
+        DICOM with header, e.g., as read by pydicom.dcmread.
+    da_tag: str
+      Dicom tag with DA value representation
+    tm_tag: str
+      Dicom tag with TM value representation
+    """
+    # https://dicom.nema.org/medical/dicom/current/output/chtml/part05/sect_6.2.html
+    date_str = dcm_data[da_tag].value.strip()
+    fmts = ["%Y%m%d",]
+    date = strptime(date_str, fmts)
+
+    time_str = dcm_data[tm_tag].value.strip()
+    fmts = ["%H", "%H%M", "%H%M%S", "%H%M%S.%f"]
+    time = strptime(time_str, fmts)
+
+    datetime_obj = datetime.datetime.combine(date.date(), time.time())
+
+    if utc_offset_dcm := dcm_data.get((0x0008, 0x0201)):
+        utc_offset = utc_offset_dcm.value.strip()
+        datetime_obj = datetime_utc_offset(datetime_obj, utc_offset) if utc_offset else datetime_obj
+    return datetime_obj
+
+
+def strptime_dcm_dt(dcm_data: dcm.Dataset, dt_tag: TagType) -> datetime.datetime:
+    r"""
```

Comment:
I know that you might have followed a trend of some docstrings sneaking in with `r` for no reason, but if no reason here in new ones -- please do not use them since adds cognitive burden to later possibly look for something needing this `raw` annotation